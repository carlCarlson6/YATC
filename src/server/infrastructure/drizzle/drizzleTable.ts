import { pgTableCreator } from 'drizzle-orm/pg-core'

export const drizzleTable = pgTableCreator((name) => `yatc_${name}`);