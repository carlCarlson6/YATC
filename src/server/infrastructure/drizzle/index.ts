import * as baseSchema from "src/server/infrastructure/drizzle/base.drizzle.schema";
import * as followsSchema from "src/server/user/follows/follows.drizzle.schema"
import * as emojisSchema from "src/server/publish-emojeet/emojis.drizzle.schema";
import { sql } from '@vercel/postgres';
//import { drizzle } from 'drizzle-orm/vercel-postgres';
import { drizzle } from 'drizzle-orm/postgres-js';

import postgres from 'postgres';
import { env } from "src/env.mjs";


export const drizzleSchema = {
	...baseSchema,
  ...followsSchema,
  ...emojisSchema,
};

//export const drizzleDb = drizzle(sql, { schema: drizzleSchema } );

export const drizzleDb = drizzle(postgres(env.NEON_DB), { schema: drizzleSchema } );
export type DrizzleDb = typeof drizzleDb;