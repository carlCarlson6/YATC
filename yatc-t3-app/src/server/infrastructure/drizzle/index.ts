import { Client } from "@planetscale/database";
import { drizzle } from "drizzle-orm/planetscale-serverless";
import { env } from "yact/env.mjs";
import * as baseSchema from "yact/server/infrastructure/drizzle/base.drizzle.schema";
import * as tweetSchema from "yact/server/send-tweet/tweet.drizzle.schema";
import * as followsSchema from "yact/server/user/follow/follow.drizzle.schema"

const schema = {
	...baseSchema,
  ...tweetSchema,
  ...followsSchema,
};

export const drizzleDb = drizzle(
  new Client({
    url: env.DATABASE_URL,
  }).connection(),
  { schema }
);

export type DrizzleDb = typeof drizzleDb;