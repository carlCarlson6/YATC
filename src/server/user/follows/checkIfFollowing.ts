import { eq } from "drizzle-orm";
import { type DrizzleDb } from "src/server/infrastructure/drizzle";
import { followsTable } from "./follow.drizzle.schema";

export const checkIfFollowingWithDrizzle = (db: DrizzleDb) => async (
  userId: string, 
  userIsFollowingId: string
) => {
  const results = await db
    .select({})
    .from(followsTable)
    .where(eq(followsTable.userId, userId) && eq(followsTable.isFollowingUserId, userIsFollowingId))
    .execute();
  return results.length > 0;
};

export type CheckIfFollowing = ReturnType<typeof checkIfFollowingWithDrizzle>;