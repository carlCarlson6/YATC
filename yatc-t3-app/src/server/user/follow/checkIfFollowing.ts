import { eq } from "drizzle-orm";
import { DrizzleDb, drizzleDb } from "yact/server/infrastructure/drizzle";
import { follows } from "./follow.drizzle.schema";

export const checkIfFollowingWithDrizzle = (db: DrizzleDb) => async (
  userWhoIsFollowing: string, 
  userWhoIsFollowed: string
) => {
  const result = await db.query.follows.findFirst({
    where: eq(follows.userWhoIsFollowing, userWhoIsFollowing) && eq(follows.userWhoIsFollowed, userWhoIsFollowed)
  }).execute();
  return !result ? false : true;
};

export const checkIfFollowing = checkIfFollowingWithDrizzle(drizzleDb);

export type CheckIfFollowing = typeof checkIfFollowing;