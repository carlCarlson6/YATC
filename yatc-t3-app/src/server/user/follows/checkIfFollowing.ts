import { eq } from "drizzle-orm";
import { type DrizzleDb } from "src/server/infrastructure/drizzle";
import { followsTable } from "./follow.drizzle.schema";
import { tweetsTable } from "src/server/publish-tweet/tweet.drizzle.schema";

export const checkIfFollowingWithDrizzle = (db: DrizzleDb) => async (
  userId: string, 
  userIsFollowingId: string
) => {
  const results = await db
    .select({})
    .from(followsTable)
    .where(eq(followsTable.userId, userId) && eq(followsTable.isFollowingUserId, userIsFollowingId))
    .execute();
  const result = results.at(0);
  return !result ? false : true;
};

export type CheckIfFollowing = ReturnType<typeof checkIfFollowingWithDrizzle>;