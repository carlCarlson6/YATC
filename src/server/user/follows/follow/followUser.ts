"use server"

import { drizzleDb, type DrizzleDb } from "src/server/infrastructure/drizzle";
import { followsTable } from "../follows.drizzle.schema";
import { checkIfFollowingWithDrizzle, type CheckIfFollowing } from "../checkIfFollowing";
import { randomUUID } from "crypto";
import { validateAuth } from "src/server/validateAuth";
import { z } from "zod";

const followUserInputSchema = z.object({
  userToFollow: z.string().min(1),
});

const followUser = ({ checkIfFollowing, storeFollower }: {
  checkIfFollowing: CheckIfFollowing,
  storeFollower: (userWhoFollos: string, userToFollow: string) => Promise<void>,
}) => async (input: {
  userToFollow: string,
}) => {
  const user = await validateAuth();
  const {userToFollow} = await followUserInputSchema.parseAsync(input);

  const isAlreadyFollowing = await checkIfFollowing(user.id, userToFollow);
  if (isAlreadyFollowing) {
    console.info("already followed - aborting operation");
    return;
  }
  
  await storeFollower(user.id, userToFollow);
};

const storeFollowerOnDrizzle = (db: DrizzleDb) => async (
  userId: string, 
  userToFollowId: string
) => {
  await db.insert(followsTable).values({
    id: randomUUID(),
    userId: userId,
    isFollowingUserId: userToFollowId,
  }).execute();
}

export default followUser({
  checkIfFollowing: checkIfFollowingWithDrizzle(drizzleDb),
  storeFollower: storeFollowerOnDrizzle(drizzleDb),
});