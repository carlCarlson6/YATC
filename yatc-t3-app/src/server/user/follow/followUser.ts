import { type DrizzleDb } from "yact/server/infrastructure/drizzle";
import { protectedProcedure } from "yact/server/infrastructure/trpc";
import z from "zod";
import { follows } from "./follow.drizzle.schema";
import { CheckIfFollowing, checkIfFollowingWithDrizzle } from "./checkIfFollowing";

export const followUserProcedure = protectedProcedure
  .input(z.object({
    userToFollow: z.string().min(1)
  }))
  .mutation(({ctx: {db, session: {user}}, input}) => followUser({
    checkIfFollowing: checkIfFollowingWithDrizzle(db),
    storeFollower:    storeFollowerOnDrizzle(db),
  })(user.id, input.userToFollow));

const followUser = ({ checkIfFollowing, storeFollower }: {
  checkIfFollowing: CheckIfFollowing,
  storeFollower: (userWhoFollos: string, userToFollow: string) => Promise<void>,
}) => async (
  userWhoFollos: string,
  userToFollow: string
) => {
  const isAlreadyFollowing = await checkIfFollowing(userWhoFollos, userToFollow);
  if (isAlreadyFollowing) {
    console.info("already followed - aborting operation");
    return;
  }
  await storeFollower(userWhoFollos, userToFollow);
};

const storeFollowerOnDrizzle = (db: DrizzleDb) => async (userWhoFollos: string, userToFollow: string) => {
  await db.insert(follows).values({
    userWhoIsFollowed: userToFollow,
    userWhoIsFollowing: userWhoFollos,
  }).execute();
}