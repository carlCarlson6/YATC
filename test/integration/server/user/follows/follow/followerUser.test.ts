import { expect, test } from 'vitest';
import { checkIfFollowingWithDrizzle } from 'src/server/user/follows/checkIfFollowing';
import { setupDockerTestDb } from 'test/helpers/setupTestingDb';
import { followsTable } from 'src/server/user/follows/follows.drizzle.schema';
import { followUser, storeFollowerOnDrizzle } from 'src/server/user/follows/follow/followUser';
import { mockAuthValidator } from '../../../../../helpers/mockAuthValidator';

const userId = "some-user-id";
const user = {
  id: userId,
  name: "some-user-name",
  avatar: "some-avatar"
}

test("GivenUser_WhenFollowAnotherUser_ThenFollowingCountIncreases", async () => {
  const userToFollow = "some-user-id-who-is-followed";
  const {db} = await setupDockerTestDb();

  await followUser({
    checkIfFollowing: checkIfFollowingWithDrizzle(db),
    storeFollower: storeFollowerOnDrizzle(db),
    auth: mockAuthValidator(user)
  })({userToFollow});

  const count = (await db.query.followsTable.findMany().execute()).length;
  expect(count).toBe(1);
});

test("GivenUser_WhenFollowAnotherUser_ThenAFollowIsStored", async () => {
  const userToFollow = "some-user-id-who-is-followed";
  const {db} = await setupDockerTestDb();

  await followUser({
    checkIfFollowing: checkIfFollowingWithDrizzle(db),
    storeFollower: storeFollowerOnDrizzle(db),
    auth: mockAuthValidator(user)
  })({userToFollow});

  const follow = await db.select({ userId: followsTable.userId, whoIsFollowed: followsTable.isFollowingUserId })
    .from(followsTable)
    .execute();
  expect(follow).toMatchSnapshot();
});