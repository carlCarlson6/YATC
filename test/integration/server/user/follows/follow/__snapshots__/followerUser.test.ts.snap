// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GivenUser_WhenFollowAnotherUser_ThenAFollowIsStored 1`] = `
[
  {
    "userId": "some-user-id",
    "whoIsFollowed": "some-user-id-who-is-followed",
  },
]
`;
